### PR TITLE
Don't require Tweedledum for `qiskit.circuit.classicalfunction` on import

### PR DESCRIPTION
**Pull request by Eric-Arellano**
### Summary

Per https://github.com/Qiskit/qiskit-terra/pull/8947, tweedledum has issues like poor macOS support.

`qiskit.circuit.classicalfunction` was requiring Tweedledum to be installed at import time. That contributes to issues with building docs on M1s because [autodoc](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html) works by importing the module:

> For Sphinx (actually, the Python interpreter that executes Sphinx) to find your module, it must be importable. 

Instead, we can make the Tweedledum errors less eager. Only error when certain functions or classes that need Tweedledum are used.

### Details and comments

Before:

```
>>> from qiskit.circuit import classicalfunction
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/arellano/code/qiskit-terra/qiskit/circuit/classicalfunction/__init__.py", line 96, in <module>
    HAS_TWEEDLEDUM.require_now("classical function oracles")
  File "/Users/arellano/code/qiskit-terra/qiskit/utils/lazy_tester.py", line 223, in require_now
    raise MissingOptionalLibraryError(
qiskit.exceptions.MissingOptionalLibraryError: "The 'tweedledum' library is required to use 'classical function oracles'. You can install it with 'pip install tweedledum'."
```

After:

```
>>> from qiskit.circuit import classicalfunction
>>> from qiskit.circuit.classicalfunction import utils
```

(I'm still having issues building the docs on my M1, but for different reasons.)

### CODE REVIEW

File: qiskit/circuit/classicalfunction/classicalfunction.py [hunk 1]
```diff
@@ -96,6 +94,7 @@ def types(self):
             ret.append({k: v[0] for k, v in scope.items()})
         return ret
 
+    @HAS_TWEEDLEDUM.require_in_call
```
Comment (jakelishman) [inline qiskit/circuit/classicalfunction/classicalfunction.py:97]: It's probably best to stick make this `require_in_instance` as well tbh - you can't do anything meaningful with the class at all without tweedledum.

**Review by jakelishman — COMMENTED** [review]
This looks generally good to me.  You might consider if it's worth putting in a manual message to each of the `require` calls to explain what functionality is missing, but I don't know if it'll make things too much clearer - I think the machinery should try and infer something vaguely sensible if the message isn't supplied.

**Comment by Eric-Arellano** [comment]
> I think the machinery should try and infer something vaguely sensible if the message isn't supplied.

Yeah, the decorator docstrings say they will explain that the class is not usable for `@require_in_instance`. For `@require_in_call`, it will use the function name.

**Review by jakelishman — APPROVED** [review]
Looks good, thanks!

We can probably remove `tweedledum` from `requirements-dev.txt` after this, but we can follow up in a separate PR.

The reason I brought up the message is that I think sometimes these classes are used implicitly (from the user's perspective) so just giving the class name might not entirely help, but I don't really remember where if anywhere, so it's probably not important.

**Pull request merged**